client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala

```diff
@@ -79,14 +85,19 @@ class ChangePartitionManager(
                 batchHandleChangePartitionExecutors.submit {
                   new Runnable {
                     override def run(): Unit = {
-                      val distinctPartitions = requests.synchronized {
-                        // For each partition only need handle one request
-                        requests.asScala.filter { case (partitionId, _) =>
-                          !inBatchPartitions.get(shuffleId).contains(partitionId)
-                        }.map { case (partitionId, request) =>
-                          inBatchPartitions.get(shuffleId).add(partitionId)
-                          request.asScala.toArray.maxBy(_.epoch)
-                        }.toArray
+                      val distinctPartitions = {
+                        val requestSet = inBatchPartitions.get(shuffleId)
+                        val locksForShuffle = locks.computeIfAbsent(shuffleId, locksRegisterFunc)
+                        requests.asScala.map { case (partitionId, request) =>
+                          locksForShuffle(partitionId % locksForShuffle.length).synchronized {
+                            if (!requestSet.contains(partitionId)) {
+                              requestSet.add(partitionId)
```

Comment:
Adding a check here can avoid replying to the same partition locations to the client twice.
`requests.containsKey(partitionId)`

Reply:
done